### PR TITLE
porting Cluster heartbeat timings, hardened Akka.Cluster serialization

### PR DESCRIPTION
**Pull request by Aaronontheweb**
port https://github.com/akka/akka/pull/27281
port https://github.com/akka/akka/pull/25183
port https://github.com/akka/akka/pull/24625

This PR introduces three changes:

1. Migrates the Akka.Cluster `ClusterMessageSerializer` to use `SerializerWithStringManifest` in order to avoid the wire format being broken by future type changes - all of the current type names are used as the string manifests to preserve backwards-compatible wire compatibility for now.
2. Introduces the wire format changes for measuring heartbeat latency, which can be useful for troubleshooting reachability issues inside a cluster. However, these wire format changes can be deserialized starting with Akka.NET v1.4.19 but none of these new message types will be emitted until at least Akka.NET v1.5 since introducing the change now would break wire compatibility for current Akka.Cluster  users.

### CODE REVIEW

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt [hunk 1]
```diff
@@ -437,11 +437,12 @@ namespace Akka.Cluster.SBR
 }
 namespace Akka.Cluster.Serialization
 {
-    public class ClusterMessageSerializer : Akka.Serialization.Serializer
+    [Akka.Annotations.InternalApiAttribute()]
+    public class ClusterMessageSerializer : Akka.Serialization.SerializerWithStringManifest
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt:441]: So even though this is part of the public API, I don't think this is much of a "breaking change" as the serializer isn't called directly from user code (that I'm aware of) and it's really meant to be an internal api.

**Comment by Aaronontheweb** [comment]
Applying "No Merge" to this for now - put together a cluster with some v1.4.18 nodes and some nodes running this code and there appears to be a logging / serialization issue that causes new nodes to become unreachable when they join the cluster. Going to continue to investigate that.

**Comment by Aaronontheweb** [comment]
Found this in the Linux test suite

```
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateISerializableItem(JToken token, Type type, JsonISerializableContract contract, JsonProperty member)
   at Newtonsoft.Json.Serialization.JsonFormatterConverter.Convert(Object value, Type type)
   at System.Runtime.Serialization.SerializationInfo.GetValue(String name, Type type)
   at System.Exception..ctor(SerializationInfo info, StreamingContext context)
   at System.AggregateException..ctor(SerializationInfo info, StreamingContext context)
   at lambda_method(Closure , Object[] )
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateISerializable(JsonReader reader, JsonISerializableContract contract, JsonProperty member, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Akka.Serialization.NewtonSoftJsonSerializer.FromBinary(Byte[] bytes, Type type) in /home/vsts/work/1/s/src/core/Akka/Serialization/NewtonSoftJsonSerializer.cs:line 245
   at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest) in /home/vsts/work/1/s/src/core/Akka/Serialization/Serialization.cs:line 436
   at Akka.Actor.ActorCell.SerializeAndDeserializePayload(Object obj) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 547
   at Akka.Actor.ActorCell.SerializeAndDeserialize(Envelope envelope) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 524
   at Akka.Actor.ActorCell.SendMessage(Envelope message) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 407
  Passed Akka.Cluster.Tests.ClusterSpec.A_cluster_must_use_the_address_of_the_remote_transport [129 ms]
[WARNING][04/13/2021 00:47:57][Thread 0055][akka://AkkaSpec/deadLetters] DeadLetter from [akka://AkkaSpec/system/cluster/core/daemon#340158463] to [akka://AkkaSpec/deadLetters]: <Received dead letter from [akka://AkkaSpec/system/cluster/core/daemon#340158463]: Akka.Cluster.InternalClusterAction+Join: UniqueAddress: (akka.tcp://AkkaSpec@0.0.0.0:34631, 1993954935) wants to join on Roles []>
[WARNING][04/13/2021 00:47:57][Thread 0055][akka://AkkaSpec/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FAkkaSpec%40127.0.0.1%3A9999-1/endpointWriter] DeadLetter from [akka://AkkaSpec/deadLetters] to [akka://AkkaSpec/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FAkkaSpec%40127.0.0.1%3A9999-1/endpointWriter#1550700525]: <Received dead letter from [akka://AkkaSpec/deadLetters]: Failure: Akka.Remote.Transport.Akk
```

**Comment by Aaronontheweb** [comment]
> Applying "No Merge" to this for now - put together a cluster with some v1.4.18 nodes and some nodes running this code and there appears to be a logging / serialization issue that causes new nodes to become unreachable when they join the cluster. Going to continue to investigate that.

False alarm - was a bug with my Kubernetes service definition I used to connect the two services together. No wire problems at all between 1.4.19-beta* and 1.4.18 nodes in a cluster.

**Comment by Aaronontheweb** [comment]
Ah, this is the spec that failed:

```
  Failed Akka.Cluster.Tests.ClusterLogVerboseEnabledSpec.A_cluster_must_log_verbose_cluster_events_when_log_info_verbose_is_on [4 s]
  Error Message:
   Timeout (00:00:03) while waiting for messages. Only received 0/1 messages that matched filter [Info when Message contains " - event MemberUp"]
```

**Comment by Aaronontheweb** [comment]
> Found this in the Linux test suite
> 
> ```
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateISerializableItem(JToken token, Type type, JsonISerializableContract contract, JsonProperty member)
>    at Newtonsoft.Json.Serialization.JsonFormatterConverter.Convert(Object value, Type type)
>    at System.Runtime.Serialization.SerializationInfo.GetValue(String name, Type type)
>    at System.Exception..ctor(SerializationInfo info, StreamingContext context)
>    at System.AggregateException..ctor(SerializationInfo info, StreamingContext context)
>    at lambda_method(Closure , Object[] )
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateISerializable(JsonReader reader, JsonISerializableContract contract, JsonProperty member, String id)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ResolvePropertyAndCreatorValues(JsonObjectContract contract, JsonProperty containerProperty, JsonReader reader, Type objectType)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
>    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
>    at Akka.Serialization.NewtonSoftJsonSerializer.FromBinary(Byte[] bytes, Type type) in /home/vsts/work/1/s/src/core/Akka/Serialization/NewtonSoftJsonSerializer.cs:line 245
>    at Akka.Serialization.Serialization.Deserialize(Byte[] bytes, Int32 serializerId, String manifest) in /home/vsts/work/1/s/src/core/Akka/Serialization/Serialization.cs:line 436
>    at Akka.Actor.ActorCell.SerializeAndDeserializePayload(Object obj) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 547
>    at Akka.Actor.ActorCell.SerializeAndDeserialize(Envelope envelope) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 524
>    at Akka.Actor.ActorCell.SendMessage(Envelope message) in /home/vsts/work/1/s/src/core/Akka/Actor/ActorCell.cs:line 407
>   Passed Akka.Cluster.Tests.ClusterSpec.A_cluster_must_use_the_address_of_the_remote_transport [129 ms]
> [WARNING][04/13/2021 00:47:57][Thread 0055][akka://AkkaSpec/deadLetters] DeadLetter from [akka://AkkaSpec/system/cluster/core/daemon#340158463] to [akka://AkkaSpec/deadLetters]: <Received dead letter from [akka://AkkaSpec/system/cluster/core/daemon#340158463]: Akka.Cluster.InternalClusterAction+Join: UniqueAddress: (akka.tcp://AkkaSpec@0.0.0.0:34631, 1993954935) wants to join on Roles []>
> [WARNING][04/13/2021 00:47:57][Thread 0055][akka://AkkaSpec/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FAkkaSpec%40127.0.0.1%3A9999-1/endpointWriter] DeadLetter from [akka://AkkaSpec/deadLetters] to [akka://AkkaSpec/system/endpointManager/reliableEndpointWriter-akka.tcp%3A%2F%2FAkkaSpec%40127.0.0.1%3A9999-1/endpointWriter#1550700525]: <Received dead letter from [akka://AkkaSpec/deadLetters]: Failure: Akka.Remote.Transport.Akk
> ```

This is still apparently a `serialize-messages = on` issue - I can see it in several of the `ClusterSpec` tests, which still pass despite issues with this.

**Review by ismaelhamed — APPROVED** [review]
LGTM

**Comment by Aaronontheweb** [comment]
This should also help with https://github.com/akkadotnet/akka.net/issues/4849

**Pull request merged**